rfcs/0044-node-status-and-node-error-backend.md

```diff
@@ -0,0 +1,61 @@
+## Summary
+
+[summary]: #summary
+
+This RFC proposes that we use AWS stack to be the backend of the node status/error system. To be specific by AWS stack, I mean we would use S3 as the storage backend; OpenSearch as the search engine; Kibana as the visualization and charting tool.
+
+## Motivation
+
+[motivation]:#motivation
+
+We need a backend for node staus/error systems. Candidates for the backend include AWS stack and Grafana Loki and LogDNA.
+
+## Implementation
```

Comment:
I think this is still missing details on how we're going to meet some of the requirements in the PRD for node status and node error. Maybe you could argue that this RFC is scoped to the "backend", but actually I think in order to understand the backend it makes sense to holistically think of the end-to-end flow of the data from the "frontend" too. Can we include this information in this RFC?

Here are some that I think need to be referenced/explained (in addition to what I mentioned in the below comments):

1. How will the Mina client handle the status/error opt-in for the data collection?
2. The frequency of the data sending
3. How can the user specify the destination of the logs? (if they want to run their own collection infrastructure)
4.  What are the mechanisms for surfacing all the data? Should we reuse the code we have for prometheus metrics or should it be different?
5. Where are the places that will we add the exception collection hook to send errors? How will the errors be indexed and queried that we're able to learn from the issues in our system?

Reply:
1. In the current implementation, there would be a cli parameter for the destination url of the node status report. If the user provides that then the node status system would send reports there. Same applies to the node error system. I think we need to do some change there, since now our backend is just the kenesis firehose endpoint. So there would be no url. I would image for this design there would be 2 cli arguments. One required parameter to trigger the node status/error system; one optional parameter to specify the destination if user wants to send it to somewhere else.

2. For the node status system, it sends the status report every 5 slots. For the node error system, it would send a report before it crashes.

3. Already answered in 1.

4. In the implementation, I actually reused a lot of prometheus metrics.

5. For the node error system, I was adding the hook to the `handle_crash` function. The same function was used to generate the current crash reports for mina nodes. For more about the node error collection system, you can take a look at this RFC: https://github.com/MinaProtocol/mina/pull/9526. I failed to follow the exact requirements, for example, I failed to extract locations of exceptions. And the backtrace stack is considered part of the exception. Basically I am just reporting the exception from mina. Do you have any suggestion that I can do better than that?

Reply:
As mentioned in Slack, can you add these answers to the actual document itself so we have a record of what our decisions ended up being?

1. I guess we can provide a bucket name or something for the kinesis firehose? I think we should do it with one CLI argument like `--share-report-data <destination-bucket>` or whatever name we use for kinesis storage. Then the boolean on/off and the destination are in the same command. I think we shouldn't hardcode our destination by default -- but try to name the kinesis endpoint "mina-report-data" or something
2. Sounds good
3. Sounds good
4. Can you mention here how the integration works with the Prometheus metrics?
5. Oh I missed the other RFC, I think we should combine them actually. Sorry I missed the original one -- a few changes: (1) it should be off by default (see the requirements doc https://www.notion.so/minaprotocol/Node-Error-Collection-030f806ede5345beb0089ffc8666f238 ) and (2) we need to adjust it to point to kinesis now and not a REST URL, right? I think it's important to evaluate the whole thing holistically.

Reply:
1. I think providing a destination URL would be more flexible than using a `destination-bucket`. Let's say we want to switch to another backend architecture in the future, then we can setup a mini-service under a sub-domain of minaprotocol.com (like what I suggested in the RFC before, that part got deleted because joseandro argued that we can push everything to the frontend which means we don't need this mini-service at all). And another thing is that we are not directly pushing to some buckets, we are pushing to the kenesis data stream. I am ok to not hardcode any default value for the destination.

4. I can find the details in this PR: https://github.com/MinaProtocol/mina/pull/9413. In short, what I did is that I extract the measurements in the metrics and filled the relevant fields in the reports with those measurements. There's no magic there.

5. Yes, by default it's off. Yes, if we decided to go with the AWS stack, then we have to change the frontend a little bit.

Reply:
It may also be helpful to report on status transitions, not just a fixed interval of 5 slots - e.g. report when status goes from sync to catchup, etc. That would allow for understanding things like network-wide bootstrap times, etc.

Reply:
Great -- using a URL sounds good to me 👍 .

@jrwashburn that's a good idea -- @ghost-not-in-the-shell would it be hard to capture those status transitions?

Reply:
Sync status is part of the "node status report". If we are interested in the bootstrap time, then we could add that to the status report directly. Personally I think sending reports every 5 slots is enough for the status report system.